src/Gaffer/Animation.cpp

```diff
@@ -38,93 +38,1892 @@
+						s = sl + ds;
+					}
+					else
+					{
+						s -= ds;
+					}
+				}
+
+				assert( s >= sl );
+				assert( s <= sh );
+
+				// check for convergence
+
+				if( std::abs( ds ) < std::numeric_limits< double >::epsilon() )
+				{
+					break;
+				}
+			}
+
+			return s;
+		}
+	};
+
+} // namespace
+
+namespace Gaffer
+{
+
+//////////////////////////////////////////////////////////////////////////
+// Time implementation
+//////////////////////////////////////////////////////////////////////////
+
+Animation::Time::Time()
+: m_ticks( static_cast< std::int64_t >( 0 ) )
+{}
+
+Animation::Time::Time( const std::int64_t ticks )
+: m_ticks( ticks )
+{}
+
+Animation::Time::Time( const double value, const Animation::Time::Units units )
+: m_ticks( static_cast< std::int64_t >( std::round( value * (
+	static_cast< double >( Units::Ticks ) /
+	static_cast< double >( units ) ) ) ) )
+{}
+
+Animation::Time::Time( const double value, const double units )
+: m_ticks( ( units > 0.0 )
+	? static_cast< std::int64_t >( std::round( value * (
+		static_cast< double >( Units::Ticks ) /
+		std::min( units, static_cast< double >( Units::Ticks ) ) ) ) )
+	: static_cast< std::int64_t >( 0 ) )
+{}
+
+Animation::Time::Time( Animation::Time const& rhs )
+: m_ticks( rhs.m_ticks )
+{}
+
+Animation::Time::~Time()
+{}
+
+Animation::Time& Animation::Time::operator  = ( const Animation::Time& rhs )
+{
+	m_ticks = rhs.m_ticks;
+	return *this;
+}
+
+Animation::Time& Animation::Time::operator += ( const Animation::Time& rhs )
+{
+	m_ticks += rhs.m_ticks;
+	return *this;
+}
+
+Animation::Time& Animation::Time::operator -= ( const Animation::Time& rhs )
+{
+	m_ticks -= rhs.m_ticks;
+	return *this;
+}
+
+Animation::Time& Animation::Time::operator /= ( const Animation::Time& rhs )
+{
+	m_ticks /= rhs.m_ticks;
+	return *this;
+}
+
+Animation::Time& Animation::Time::operator %= ( const Animation::Time& rhs )
+{
+	m_ticks = std::labs( m_ticks % rhs.m_ticks );
+	return *this;
+}
+
+std::int64_t Animation::Time::getTicks() const
+{
+	return m_ticks;
+}
+
+double Animation::Time::getReal( const Animation::Time::Units units ) const
+{
+	return (
+		static_cast< double >( m_ticks ) / (
+		static_cast< double >( Units::Ticks ) / static_cast< double >( units ) ) );
+}
+
+double Animation::Time::getReal( const double units ) const
+{
+	return ( units > 0.0 )
+		? ( static_cast< double >( m_ticks ) / (
+			static_cast< double >( Units::Ticks ) / units ) )
+		: static_cast< double >( 0 );
+}
+
+double Animation::Time::getSeconds() const
+{
+	return getReal( Units::Seconds );
+}
+
+void Animation::Time::snap( const double units )
+{
+	if( units > 0.0 )
+	{
+		// NOTE : round to nearest unit then convert back to ticks and round to nearest tick
+
+		const double s = static_cast< double >( Units::Ticks ) /
+			std::min( units, static_cast< double >( Units::Ticks ) );
+
+		m_ticks = static_cast< std::int64_t >( std::round( std::round( static_cast< double >( m_ticks ) / s ) * s ) );
```

Comment:
the conversion to double is exact for ticks up to 2^53 which gives roughly +/-147 days of animation or +/-5,000,000 frames at 24 frames a second, I suppose this could be a problem if people are using frame numbers over 5,000,000 ...